### PR TITLE
Fix `to_unit` with scale exceeding int64_max

### PR DESCRIPTION
**Pull request by SimonHeybrock**
This was introduced by #2607. No released version is affected.

### CODE REVIEW

File: lib/variable/test/to_unit_test.cpp [hunk 1]
```diff
@@ -274,3 +274,19 @@ TEST(ToUnitTest, small_to_large_rounding_error_double) {
...
             one_m);
 }
+
+TEST(ToUnitTest, small_number_to_small_unit) {
+  const auto area = units::angstrom * units::angstrom;
+  const auto small = makeVariable<double>(units::Unit("m**2"), Values{1e-20});
+  const auto result = to_unit(small, area);
+  EXPECT_EQ(result.unit(), area);
+  EXPECT_DOUBLE_EQ(result.value<double>(), 1.0);
+}
```
Comment (jl-wynen) [inline lib/variable/test/to_unit_test.cpp:284]: Did you check if this test passes without going through `base_scale`? I'm wondering how much if anything we are gaining from rounding.

File: lib/variable/test/to_unit_test.cpp [hunk 2]
```diff
@@ -274,3 +274,19 @@ TEST(ToUnitTest, small_to_large_rounding_error_double) {
...
+
+TEST(ToUnitTest, small_number_to_small_unit_non_power_of_10) {
+  const auto area = units::Unit("1.45e-21");
```
Comment (nvaytet) [inline lib/variable/test/to_unit_test.cpp:287]: `area` name is a bit confusing here.

File: lib/variable/test/to_unit_test.cpp [hunk 3]
```diff
@@ -274,3 +274,19 @@ TEST(ToUnitTest, small_to_large_rounding_error_double) {
                     units::Unit("m")),
             one_m);
 }
+
+TEST(ToUnitTest, small_number_to_small_unit) {
```
Comment (nvaytet) [inline lib/variable/test/to_unit_test.cpp:278]: Did the problem (running out of range with int64) arise only with small number or also with large numbers?

File: lib/variable/test/to_unit_test.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply lib/variable/test/to_unit_test.cpp:284]: Well, probably that is two different questions:
- Even with a scale of 1e9 we gain by going through `base_scale`.
- This test in particular is essentially a safeguard against massive issues, not rounding errors, so I don't know if `EXPECT_DOUBLE_EQ` is sensitive to those rounding errors.

Comment (SimonHeybrock) [reply lib/variable/test/to_unit_test.cpp:284]: But to answer your actual question: I don't think rounding does anything in many of these cases. But this seemed to be the solution with the simplest branching logic (remember our chat, where I had temporarily tried putting a limit on `scale`).

**Review by jl-wynen — APPROVED** [review]
I am not convinced that this is free of nasty surprises. But it is an improvement. And I don't see a good way of verifying it other than trying lot of cases.

**Comment by SimonHeybrock** [comment]
> I am not convinced that this is free of nasty surprises. But it is an improvement. And I don't see a good way of verifying it other than trying lot of cases.

Is this something a `hypothesis` test could deal with?

**Comment by jl-wynen** [comment]
> Is this something a hypothesis test could deal with?

Only if you can verify that the result is correct. Since we cannot rely on float or regular int operations to be applicable, we cannot use them to compute the 'expected' result of a test. Or do you have an idea?

**Comment by SimonHeybrock** [comment]
@jl-wynen Have a look at the latest change. Does that make sense, or is that missing the point?

File: tests/hypothesis/to_unit_test.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+# SPDX-License-Identifier: BSD-3-Clause
```
Comment (nvaytet) [inline tests/hypothesis/to_unit_test.py:1]: Did you try to see if you got any failures in these tests if you revert your changes?

Comment (SimonHeybrock) [reply tests/hypothesis/to_unit_test.py:1]: Yes, lots of errors.

File: tests/hypothesis/to_unit_test.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+# @author Jan-Lukas Wynen
+import pytest
+from hypothesis import given
+from hypothesis import strategies as st
+
+import scipp as sc
+
+
+@given(st.floats(min_value=1e-300, max_value=1e300))
+def test_to_unit_value_1(x):
```
Comment (jl-wynen) [inline tests/hypothesis/to_unit_test.py:12]: Why 'value_1' ?

File: tests/hypothesis/to_unit_test.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    unit = sc.Unit(f'{x:.12E}')
+    var = sc.scalar(1.23 * x, unit='')
+    result = sc.to_unit(var, unit)
+    assert result.unit == unit
+    assert pytest.approx(result.value, 1e-14) == 1.23
```
Comment (jl-wynen) [inline tests/hypothesis/to_unit_test.py:17]: Why 1e-14? Given that `unit` is only encoded to 13 digits, I would not expect the test to work with this tolerance. But apparently it does?

File: tests/hypothesis/to_unit_test.py (hunk 5 above)
Comment (SimonHeybrock) [reply tests/hypothesis/to_unit_test.py:12]: Ah, but I forgot to remove the digits...

File: tests/hypothesis/to_unit_test.py (hunk 6 above)
Comment (SimonHeybrock) [reply tests/hypothesis/to_unit_test.py:17]: 🤦 while `pytest.approx` has a default of `abs=None`, it uses `1e-12` as default absolute error bound

File: tests/hypothesis/to_unit_test.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2022 Scipp contributors (https://github.com/scipp)
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply tests/hypothesis/to_unit_test.py:2]: Done.

**Review by jl-wynen — APPROVED** [review]

**Pull request merged**